src/differentiation/compute_hessian_ad.jl

```diff
@@ -0,0 +1,105 @@
+struct ForwardColorHesCache{THS, THC, TI<:Integer, TD, TGF, TGC, TG}
+    sparsity::THS
+    colors::THC
+    ncolors::TI
+    D::TD
+    buffer::TD
+    grad!::TGF
+    grad_config::TGC
+    G::TG
+    dG::TG
+end
+
+
+
+function make_hessian_buffers(colorvec, x)
+    ncolors = maximum(colorvec)
+    D = hcat([float.(i .== colorvec) for i in 1:ncolors]...)
+    buffer = similar(D)
+    G = similar(x)
+    dG = similar(x)
+    return (;ncolors, D, buffer, G, dG)
+end
+
+function ForwardColorHesCache(f, 
+                              x::AbstractVector{<:Number}, 
+                              g!,
+                              colorvec::AbstractVector{<:Integer}=eachindex(x), 
+                              sparsity::Union{AbstractMatrix, Nothing}=nothing)
+    ncolors, D, buffer, G, dG = make_hessian_buffers(colorvec, x)
+    grad_config = ForwardDiff.GradientConfig(f, x)
+    
+    # If user supplied their own gradient function, make sure it has the right
+    # signature (i.e. g!(G, x) or g!(G, x, grad_config::ForwardDiff.GradientConfig))
+    if ! hasmethod(g!, (typeof(G), typeof(G), typeof(grad_config)))
+        if ! hasmethod(g!, (typeof(G), typeof(G)))
+            throw(ArgumentError("Signature of `g!` must be either `g!(G, x)` or `g!(G, x, grad_config::ForwardDiff.GradientConfig)`"))
+        end
+        # define new method that takes a GradientConfig but doesn't use it
+        g1!(G, x, grad_config) = g!(G, x)
+    else
+        g1! = g!
+    end
+    
+    if sparsity === nothing
+        sparsity = sparse(ones(length(x), length(x)))
+    end
+    return ForwardColorHesCache(sparsity, colorvec, ncolors, D, buffer, g1!, grad_config, G, dG)
+end
+
+
+function ForwardColorHesCache(f, 
+                              x::AbstractVector{<:Number},
+                              colorvec::AbstractVector{<:Integer}=eachindex(x),
+                              sparsity::Union{AbstractMatrix, Nothing}=nothing)
+    g!(G, x, grad_config) = ForwardDiff.gradient!(G, f, x, grad_config)
+    return ForwardColorHesCache(f, x, g!, colorvec, sparsity)
+end
```

Comment:
collapse this into above by making `g` an optional argument with that as its argument